toolchain/check/context.cpp

```diff
@@ -108,59 +108,166 @@ auto Context::NoteIncompleteClass(SemIR::ClassId class_id,
-    if (it == name_lookup_.end()) {
-      if (print_diagnostics) {
-        DiagnoseNameNotFound(parse_node, name_id);
+    // Look for a name in the current scope only. There are two cases where the
+    // name would be in an outer scope:
+    //
+    //  - The name is the sole component of the declared name:
+    //
+    //    class A;
+    //    fn F() {
+    //      class A;
+    //    }
+    //
+    //    In this case, the inner A is not the same class as the outer A, so
+    //    lookup should not find the outer A.
+    //
+    //  - The name is a qualifier of some larger declared name:
+    //
+    //    class A { class B; }
+    //    fn F() {
+    //      class A.B {}
+    //    }
+    //
+    //    In this case, we're not in the correct scope to define a member of
+    //    class A, so we should reject, and we achieve this by not finding the
+    //    name A from the outer scope.
+    if (auto name_it = name_lookup_.find(name_id);
+        name_it != name_lookup_.end()) {
+      auto result = name_it->second.back();
+      if (result.scope_index == current_scope_index()) {
+        return result.node_id;
       }
-      return SemIR::NodeId::BuiltinError;
     }
-    CARBON_CHECK(!it->second.empty())
+    return SemIR::NodeId::Invalid;
+  } else {
+    // TODO: Once we support `extend`, do not look into `extend`ed scopes here,
+    // following the same logic as above.
+    return LookupQualifiedName(parse_node, name_id, scope_id,
+                               /*required=*/false);
+  }
+}
+
+auto Context::LookupUnqualifiedName(Parse::Node parse_node, StringId name_id)
+    -> SemIR::NodeId {
+  // TODO: Check for shadowed lookup results.
+
+  // Find the results from enclosing lexical scopes. These will be combined with
+  // results from non-lexical scopes such as namespaces and classes.
+  llvm::ArrayRef<LexicalLookupResult> lexical_results;
+  if (auto name_it = name_lookup_.find(name_id);
+      name_it != name_lookup_.end()) {
+    lexical_results = name_it->second;
+    CARBON_CHECK(!lexical_results.empty())
         << "Should have been erased: " << strings().Get(name_id);
+  }
 
-    // TODO: Check for ambiguous lookups.
-    return it->second.back();
-  } else {
-    const auto& scope = name_scopes().Get(scope_id);
-    auto it = scope.find(name_id);
-    if (it == scope.end()) {
-      if (print_diagnostics) {
-        DiagnoseNameNotFound(parse_node, name_id);
-      }
-      return SemIR::NodeId::BuiltinError;
+  // Walk the non-lexical scopes and perform lookups into each of them.
+  for (std::size_t non_lexical_scope_index :
+       llvm::reverse(non_lexical_scope_indexes_)) {
+    auto& non_lexical_scope = scope_stack_[non_lexical_scope_index];
+
+    // If the innermost lexical result is within this non-lexical scope, then
+    // it shadows all further non-lexical results and we're done.
+    if (!lexical_results.empty() &&
+        lexical_results.back().scope_index > non_lexical_scope.index) {
```

Comment:
I'm a little confused by this. `non_lexical_scope_index` is a `size_t` instead of a `ScopeIndex`, even though it ends with `scope_index`. But you use this to get `non_lexical_scope.index`, which _is_ a `ScopeIndex`.

I was going to suggest `indexes` -> `indices`, but that's a nit. Maybe the names should be more distinct?

Maybe `non_lexical_scope_indexes_` could just maintain the set of `ScopeIndex` + `NameScopeId`, i.e. a stack of `non_lexical_scopes_`? (which would be the same size as `size_t`, right?) That would remove the need to do the `scope_stack_` lookup on 184, as a small benefit.

Reply:
Switched to a list of `ScopeIndex` + `NameScopeId` as suggested. I think we'll probably eventually also want the `scope_inst_id`, but we can add that when needed.